XC Course Question Templates

coatless opened the pull request:
Set of questions that are geared toward easy "Copying" from the XC course using the web interface editor.  Close #1121 

This PR reconfigures the example course assessments to emphasize template use. 

Assessment Hierarchy:

<img width="906" alt="Screen Shot 2020-03-19 at 11 07 52 PM" src="https://user-images.githubusercontent.com/833642/77134817-a8bda300-6a36-11ea-8fb9-d756e8326950.png">

Assessment view:

<img width="868" alt="Screen Shot 2020-03-19 at 10 54 24 PM" src="https://user-images.githubusercontent.com/833642/77134595-b45c9a00-6a35-11ea-86cc-f02ff56c9979.png">


Example questions:


<img width="828" alt="Screen Shot 2020-03-19 at 10 48 31 PM" src="https://user-images.githubusercontent.com/833642/77134627-dc4bfd80-6a35-11ea-8559-57a9fa954a48.png">
<img width="826" alt="Screen Shot 2020-03-19 at 10 48 56 PM" src="https://user-images.githubusercontent.com/833642/77134623-db1ad080-6a35-11ea-881a-a482fdd23918.png">
<img width="827" alt="Screen Shot 2020-03-19 at 10 48 45 PM" src="https://user-images.githubusercontent.com/833642/77134626-dbb36700-6a35-11ea-85ad-e678f713f33e.png">
<img width="828" alt="Screen Shot 2020-03-19 at 11 04 51 PM" src="https://user-images.githubusercontent.com/833642/77134678-12897d00-6a36-11ea-8c0d-e23e16dcf3d3.png">
<img width="828" alt="Screen Shot 2020-03-19 at 11 01 54 PM" src="https://user-images.githubusercontent.com/833642/77134622-da823a00-6a35-11ea-825f-c56aeea03f25.png">

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoice/info.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "uuid": "fd9d93ad-0658-4586-9c24-83480e4c8bb7",
+    "title": "Template pl-multiple-choice Random: Select one answer from a randomized set of responses",
```
```suggestion
    "title": "Template pl-multiple-choice with randomized question prompts from a list",
```

mwest1066 commented on exampleCourse/questions/templateFixedMultipleChoice/info.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "uuid": "2dafe1f8-7b11-4efd-8d24-77b6b457b2e9",
+    "title": "Template pl-multiple-choice Fixed: Select one answer from a fixed set of responses",
```
Can we have different types of MC question randomization:
1. One correct answer, many incorrect answers, everything hard-coded.
2. Multiple correct hard-coded answers. E.g., "Which of the following is an even number" as MC.
3. Randomization of the question prompt, selected from a list (this is the "planets" question).
4. Randomization of the question prompt using random numbers (like "demoRandomMultipleChoice").

mwest1066 replied:
```suggestion
    "title": "Template pl-multiple-choice: Fixed correct answer, randomly chosen distractors",
```

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoice/question.html:
```diff
@@ -0,0 +1,12 @@
+<pl-question-panel>
+<markdown>
+Which planet is **{{{params.custom_prompt}}}** to the sun? 
```
```suggestion
Which planet is **{{{params.question_prompt}}}** to the sun? 
```

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoice/question.html:
```diff
@@ -0,0 +1,12 @@
+</markdown>
+</pl-question-panel>
+
+<pl-multiple-choice answers-name="ans" weight="1">
```
```suggestion
<pl-multiple-choice answers-name="ans">
```

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoice/question.html:
```diff
@@ -0,0 +1,12 @@
+</pl-question-panel>
+
+<pl-multiple-choice answers-name="ans" weight="1">
+  <pl-answer correct="true"> {{{params.correct_scenario}}}</pl-answer>
```
```suggestion
  <pl-answer correct="true">{{params.correct_answer}}</pl-answer>
```

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoice/question.html:
```diff
@@ -0,0 +1,12 @@
+  <pl-answer correct="true"> {{{params.correct_scenario}}}</pl-answer>
+  <pl-answer correct="false">{{{params.wrong_scenario1}}}</pl-answer>
+  <pl-answer correct="false">{{{params.wrong_scenario2}}}</pl-answer>
+  <pl-answer correct="false">{{{params.wrong_scenario3}}}</pl-answer>
```
```suggestion
  <pl-answer correct="false">{{params.wrong_answer1}}</pl-answer>
  <pl-answer correct="false">{{params.wrong_answer2}}</pl-answer>
  <pl-answer correct="false">{{params.wrong_answer3}}</pl-answer>
```

coatless commented on exampleCourse/courseInstances/Sp15/assessments/hw06-templateQuestions/infoAssessment.json:
```diff
@@ -0,0 +1,22 @@
+    "type": "Homework",
+    "title": "PrairieLearn question templates",
+    "set": "Homework",
+    "number": "6",
```
```suggestion
    "number": "1",
```

coatless commented:
Surfacing @mwest1066 feedback in https://github.com/PrairieLearn/PrairieLearn/pull/2173#discussion_r395141939  more: 


- One correct answer, many incorrect answers, everything hard-coded.
- Multiple correct hard-coded answers. E.g., "Which of the following is an even number" as MC.
- Randomization of the question prompt, selected from a list (this is the "planets" question).
- Randomization of the question prompt using random numbers (like "demoRandomMultipleChoice").

mwest1066 commented on exampleCourse/questions/templateFixedMultipleManyAnswers/info.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "uuid": "f24e0597-bb8c-42aa-abef-6a96b6afcca8",
+    "title": "Template pl-multiple-choice with randomly chosen correct answer and distractors",
```
```suggestion
    "title": "Template pl-multiple-choice: Randomly chosen correct answer and distractors",
```

Making the title format consistent with the "Fixed correct answer" title.

coatless replied:
Argh, old data here. Done in the last commit round.

mwest1066 commented on exampleCourse/questions/templateFixedMultipleManyAnswers/question.html:
```diff
@@ -0,0 +1,20 @@
+<pl-question-panel>
+What color can be found in a rainbow?
```
Love it!

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoiceListPrompt/info.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "uuid": "fd9d93ad-0658-4586-9c24-83480e4c8bb7",
+    "title": "Template pl-multiple-choice with randomized question prompts from a list",
```
```suggestion
    "title": "Template pl-multiple-choice: Randomized question prompts from a list",
```

mwest1066 commented on exampleCourse/questions/templateCheckbox/info.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "uuid": "f555b053-5d49-4719-a850-7e4b2a3d2016",
+    "title": "Template pl-checkbox: Select one or more correct answers",
```
```suggestion
    "title": "Template pl-checkbox: Randomly chosen answers (allows multiple selection)",
```

coatless commented:
@mwest1066 I moved the question prompt generation for "planets" into the `server.py` instead of having it be split between `question.html` and `server.py`. 

I also added a numeric MC question with an example that should both decimal and integer random number generation.

mwest1066 commented on exampleCourse/questions/templateFixedMultipleChoiceManyAnswers/question.html:
```diff
@@ -0,0 +1,20 @@
+<pl-multiple-choice answers-name='colors' number-answers='4'>
+  <pl-answer correct='true'>Red</pl-answer>
+  <pl-answer correct='true'>Blue</pl-answer>
+  <pl-answer correct='true'>Purple</pl-answer>
```
```suggestion
  <pl-answer correct='true'>Violet</pl-answer>
```

Is violet/purple a UK/US thing? But see https://en.wikipedia.org/wiki/Rainbow#Number_of_colours_in_spectrum_or_rainbow

nicknytko replied:
Seem so, most people here say "purple" instead of "violet" but still learn the spectrum of the rainbow as 'ROYGBIV' -- with violet at the end.

mwest1066 commented on exampleCourse/questions/templateFixedMultipleChoiceManyAnswers/question.html:
```diff
@@ -0,0 +1,20 @@
+<pl-question-panel>
+What color can be found in a rainbow?
+</pl-question-panel>
```
```suggestion
<pl-question-panel>
<p>What color can be found in a rainbow?</p>
<p><i>This question will randomly select one of the correct answers and as many of the incorrect answers as needed to make up the full list, and will then randomly shuffle them.</i></p>
</pl-question-panel>
```

Do you think we should include brief descriptions like this?

coatless replied:
I'd omit the description... One less thing to have to delete.

mwest1066 commented on exampleCourse/courseInstances/Sp15/assessments/hw01-templateQuestions/infoAssessment.json:
```diff
@@ -0,0 +1,24 @@
+                {"id": "templateFixedMultipleChoiceSingleAnswer", "points": 1},
+                {"id": "templateFixedMultipleChoiceManyAnswers", "points": 1, "maxPoints": 2},
+                {"id": "templateRandomMultipleChoiceListPrompt", "points": 1, "maxPoints": 4},
+                {"id": "templateRandomMultipleChoiceNumericChoices", "points": 1, "maxPoints": 4}
```
```suggestion
                {"id": "templateMultipleChoiceSingleCorrect", "points": 1},
                {"id": "templateMultipleChoiceMultipleCorrect", "points": 1, "maxPoints": 2},
                {"id": "templateMultipleChoiceFromList", "points": 1, "maxPoints": 4},
                {"id": "templateMultipleChoiceNumeric", "points": 1, "maxPoints": 4}
```

I know that Fixed/Random was being used in the sense of "does the Python code do randomization", but I was worried it was misleading because even the Fixed variants are actually randomizing, just in the element itself.

coatless replied:
Updated names in new branch

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoiceListPrompt/question.html:
```diff
@@ -0,0 +1,10 @@
+<pl-question-panel>
+{{{params.question_prompt}}}
```
I actually think the previous way was easier to understand, and we should double-brace the substitution (not triple). If people want to put HTML in their substitutions then we should have a separate example for that.

coatless replied:
I aimed to use triple brackets to make the questions as robust as possible and to reduce any issues that might occur. (Who knows what will be feed in from the `server.py`?)

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoiceListPrompt/question.html:
```diff
@@ -0,0 +1,10 @@
+  <pl-answer correct="true">{{{params.correct_answer}}}</pl-answer>
+  <pl-answer correct="false">{{{params.wrong_answer1}}}</pl-answer>
+  <pl-answer correct="false">{{{params.wrong_answer2}}}</pl-answer>
+  <pl-answer correct="false">{{{params.wrong_answer3}}}</pl-answer>
```
I think we should only double-brace substitute these

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoiceNumericChoices/question.html:
```diff
@@ -0,0 +1,10 @@
+<pl-question-panel>
+If $a = {{{params.a}}}$ and $b = {{{params.b}}}$, then what would the value of $a \times b$ be? 
```
```suggestion
If $a = {{params.a}}$ and $b = {{params.b}}$, then what is $a$ multiplied by $b$? 
```

I think we should always double-brace for safety. And my eyes couldn't handle the use of a cross-product for scalar multiplication 😄

mwest1066 commented on exampleCourse/questions/templateRandomMultipleChoiceNumericChoices/server.py:
```diff
@@ -0,0 +1,24 @@
+    # Generate three distractors by changing the operation
+    data['params']['wrong_answer1'] = a * (b - 1)
+    data['params']['wrong_answer2'] = b - a
+    data['params']['wrong_answer3'] = a + b
```
This is actually a great example of how we can get accidentally generate duplicate answers. If `a = 1.2` and `b = 6` then `a * b = 7.2 = a + b`, so `wrong_answer3 = correct_answer`.

It's safer to generate distractors here with `(a - 1) * b` and `(a + 1) * b` or something like that, which must be distinct from `a * b`.

We should save this example as a good case study of how it's possible to accidentally generate bad questions.

coatless replied:
Whoops, I'll add an issue ticket on distractors.
